### ISSUE TITLE
Access Tokens?

### ISSUE THREAD

**Issue by arkuhn** [Azure-Samples/active-directory-b2c-ios-native-appauth#1]
The implementation implies authorization is successful on retrieval of access tokens; however (to my knowledge) b2c Mobile and Native apps use id tokens and refresh tokens:

https://docs.microsoft.com/en-us/azure/active-directory-b2c/active-directory-b2c-apps#mobile-and-native-apps

Shouldn't we query the authState for an ID and refresh token?

**Comment by thomfromnorth** [comment]
I've tried this sample (_which is a kind of copy and paste of appauth ios sample with azure b2c sample tenant_), even with the B2C sample tenant of Microsoft.

In my case, I just get the IdToken.
**=> The Access token is null.**

From the source code of Microsoft, I can see that they expect the access token:
`NSLog(@"Got authorization tokens. Access token: %@", authState.lastTokenResponse.accessToken);`

Moreover, I've the same result with my own tenant, I get the IdToken, but no accesstoken.

So, to summarize:
- Either there is something that I did not configure.  And if yes, this means that the documentation lacks details.
- Either Microsoft has proposed a sample without having verified the result.

Really strange.

**Comment by thomfromnorth** [comment]
By the way, Arkuhn, you're right, for your azure document.

Only IdToken (and refresh token) should be used to make call on web api.

> Allowing your iOS app to access a partner web API by using OAuth 2.0 access tokens is not currently supported. 

In my tests, I did not get the refresh token.
It's due to bad configuration (from the sample) in the scopes.
It's missing offline_access.

`new string[] { Scope.OpenId, new NSString("offline_access") }`

then I got the refresh token.

Regards,

**Comment by arkuhn** [comment]
Thanks for the feedback; I noticed the same.  

 For anyone curious, in the event you do actually need an access token (alongside the refresh/id token), it can be obtained by adding the client ID/app ID to the list of scopes.

See here:
https://docs.microsoft.com/en-us/azure/active-directory-b2c/active-directory-b2c-reference-oauth-code

**Comment by arkuhn** [comment]
I posted this issue in the comments section of the sample it came from and the creators responded to me there with the following: 

> You are correct, in this sample AAD B2C is returning an id_token.  For the purpose of securing micro services that are all part of the same logical application, that might be sufficient.  For cases where you want to secure a resource and define different scopes for each application, it would be better to use access tokens.  We are close to releasing an update that enables access tokens.  We will be updating our samples to help customers get started with the new feature once it is available.  Stay tuned!

**Comment by gsacavdm** [comment]
Hey guys, my apologies for the confusion here.

* At this point, the scope of this sample is only sign-in, NOT access to web APIs.
    * For sign-in, you only need the id_token.
    * Access tokens and refresh tokens are for access to web APIs, and thus not included in this sample.

* In general (outside of Azure AD B2C), to call web APIs, you need both access tokens and refresh tokens.
    * You get access tokens by adding the API you want to access as scope.
    * You get refresh tokens by adding the offline_access scope.
    * However, in B2C:
      * If you use the same app registration for the mobile app and the API, you can use the id_token + refresh token.
      * If you specify the client_id as a scope, you'll get an access token that's identical to the id_token, allowing you to also use access token + refresh token for app + api with the same app registration.
      * In the next few days we'll release proper support for access tokens which will allow you to have separate app registrations for the mobile app and the API, define proper scopes for the API and get a proper access token. Our next iteration of this sample will showcase this new feature.

If you can wait a few days, I'd recommend you use the more comprehensive access token approach with separate app registrations and proper scopes for the web API.

**Comment by jayababu-constient** [comment]
I'm new to Azure can u explain in detail where to add the scope to get Access token and Refresh token. 
Thanks In Advance

**Comment by thomfromnorth** [comment]
Thanks gsacavdm for the explanation.

Indeed, better to wait for your complete sample.

**Comment by arkuhn** [comment]
@jayababu-constient if you are following the sample directly, you should have a line that looks like this:

`OIDAuthorizationRequest *request = 
 [[OIDAuthorizationRequest alloc] initWithConfiguration:configuration`
                                                    `             clientId:kClientId`
                                                     `            scopes:@[OIDScopeOpenID, OIDScopeProfile]`
                                                     `            redirectURL:[NSURL URLWithString:kRedirectUri]`
                                                     `            responseType:OIDResponseTypeCode`
                                                     `            additionalParameters:nil];`

inside of the "scopes" configuration list, add different scopes you might need as follows:

`scopes:@[@"offline_access", kClientId, @"ExampleScope"]`

and so on to get the response you need. Then, accordingly, you can access them as follows:

`NSLog(@"Got authorization tokens. Access token: %@", authState.lastTokenResponse.accessToken);`
`NSLog(@"Got refresh token. Refresh token: %@", authState.lastTokenResponse.refreshToken);`
`NSLog(@"Got id token. ID token: %@", authState.lastTokenResponse.idToken);`


and so on. Hopefully this gives you a clearer idea.

**Comment by parakhj** [comment]
Although the sample currently does not acquire an access token, you can learn how to by going to 

https://docs.microsoft.com/en-us/azure/active-directory-b2c/active-directory-b2c-access-tokens

**Comment by ghost** [comment]
What the Hell is an API Scope?  How do I set up my Azure B2C Application to recognize and respond to Scopes?  I have a Native Client in WPF that never needed a scope.  Now I'm trying to convert to UWP and whatever changes were made has broken the sign in process.  The Access Token is null and it appears related to the fact that we never needed a 'Scope' with the previous version.

**Comment by RyanBuzzInteractive** [comment]
Hi there. I have an issue which seems to be related to all this. I am using basically the same implementation in my Xamarin iOS native app as the Forms demo.

I am *not* trying to access a web api. I am just using AD B2C for sign in. I do not, therefor, get an Access Token back, which makes sense.

The problem I have is that I would expect calling 'AcquireTokenSilentAsync' would return a success if we have previously signed in successfully, within the signup policy expiration window, and then try to sign in again with no internet connection.

What I find is that if I sign in, then exit the app, turn off internet, and then relaunch and try to sign in again immediately, the request just fails and I get a HTTP exception. This means users always need to have internet to launch the app.

As far as I can tell, openId and offline_access are both now implicit scopes - adding them results in an error message saying so. However, I am forced to put in a dummy scope for the auth request - it has to be dummy as I have no api to add a scope for since it's a 'sign in only' client app. This feels like a weird design, unless I am missing something.

If I have internet, it all works fine and aquire silent succeeds.

Is it the case that without an Access token (for an api) I get no offline ability? The Expires property in the AuthentictionResult is default datetime 1970 blah blah, which is supposed to be for the Access token which is null.

**Comment by sathishTkumar** [comment]
Hi,
How am to save the received id_token on sign-in? Because each time, when I kill and launch my application after successful sign-in, login screen is getting launched again.

**Comment by vikher** [comment]
is login information persisted across app launches? , the maximum time I get for Access & ID token is one day. Do I need to create my own Web API to use Refresh Token? Is there any sample on how to keep the user logged in? All I need to do I to keep the users logged in since this is a native mobile app﻿

**Comment by arkuhn** [comment]
I'm going to close this issue as the original question has been answered and the sample has been updated. If you believe you've discovered a problem, please start a new issue following the [posted guidelines. ](https://docs.microsoft.com/en-us/contribute/#create-quality-issues) There are a wealth of resources online both in and outside of the Azure docs for general OAuth token questions.
